tests/run-make/asm-incr-cache/rmake.rs

```diff
@@ -0,0 +1,16 @@
+extern crate run_make_support;
+
+use run_make_support::{rustc, tmp_dir};
+
+fn main() {
+    let inc_dir = tmp_dir();
+
+    for _ in 0..=1 {
+        rustc()
+            .input("lib.rs")
+            .arg("--emit=obj,asm,dep-info,link,mir,llvm-ir,llvm-bc")
+            .arg("--crate-type=lib")
+            .arg(format!("-Cincremental={}", inc_dir.to_str().expect("temp dir?")).as_str())
```

Comment:
Could we float the `to_str().expect()` out of the loop? Furthermore, could you change the panic message to say sth. akin to “tmpdir path isn't valid utf-8”.

Ideally, `run-make-support`'s `.arg()` would take `impl AsRef<OsStr>` instead of `&str` so we won't panic on machines where the `TMPDIR` is not valid UTF-8 (unlikely but not impossible). @jieyouxu what do you think?

Reply:
> Ideally, `run-make-support`'s `.arg()` would take `impl AsRef<OsStr>` instead of `&str` so we won't panic on machines where the `TMPDIR` is not valid UTF-8 (unlikely but not impossible).

I think that makes sense yes. But in this case, we should also provide convenience APIs to avoid annoying `OsString` constructions methinks.

So if we change `arg`, we should also change `args`

```rs
impl Rustc {
    pub fn arg<S: AsRef<OsStr>>(&mut self, val: S) -> &mut Self { ... }
    pub fn args<S: AsRef<OsStr>>(&mut self, val: &[S]) -> &mut Self { ... }
}
```

and provide a convenience method for incremental here that constructs the suitable `-Cincremental={}` with `OsString` so the test-writer doesn't have to:

```rs
impl Rustc {
    pub fn incremental<P: AsRef<Path>>(&mut self, path: P) -> &mut Self {
        let mut arg = OsString::new();
        arg.push("-Cincremental=");
        arg.push(path.as_ref())
        self.cmd.arg(&arg);
        self
    }
}
```

With this API, this test can just write

```rs
rustc()
    // lines omitted
    .incremental(&tmp_dir())
```

Reply:
> Could we float the `to_str().expect()` out of the loop? Furthermore, could you change the panic message to say sth. akin to “tmpdir path isn't valid utf-8”.

Also, with the `incremental()` API, we don't need the `expect()` or `to_str()` at all.

Reply:
Actually I realized that @pacak *did* introduce `incremental()` helper -- in another PR #122597 (mentioned in previous resolved comment <https://github.com/rust-lang/rust/pull/123221#discussion_r1546824858>). IIUC, @pacak wanted to merge this PR first? So if we could move the helper to this PR and change the `arg`/`args` signatures, that'd be great ^^

Reply:
```rust
impl Rustc {
    pub fn arg<S: AsRef<OsStr>>(&mut self, val: S) -> &mut Self { ... }
    pub fn args<S: AsRef<OsStr>>(&mut self, val: &[S]) -> &mut Self { ... }
}
```

I'm about to do some black magic to ensure this assert still works with `OsStr`
```rust
        assert!(                                                                                                                                                                                                                               
            !(["-Z", "-C"].contains(&arg) || arg.starts_with("-Z ") || arg.starts_with("-C ")),                                                                                                                                                
            "use `-Zarg` or `-Carg` over split `-Z` `arg` or `-C` `arg`"                                                                                                                                                                       
        );                                                                                   
```

Reply:
Can you just remove that assert? Just make sure the recommendation is documented on the method.